### PR TITLE
HP2640: added tape emulation

### PR DESCRIPTION
**Pull request by fulivi**
Hi & happy new year!
With this PR I'm adding the emulation of DC100 tape drives to HP264x terminals.
I extended the HTI format to handle the Manchester-encoded data of these terminals (previously HTI format only supported delta-distance modulation of HP9845,9825 & 85 systems).
Thanks.
--F.Ulivi

### CODE REVIEW

File: src/mame/drivers/hp2640.cpp [hunk 1]
```diff
@@ -1012,19 +1049,7 @@ void hp2640_base_state::cpu_mem_map(address_map &map)
...
-	map(0x8300, 0x8300).w(FUNC(hp2640_base_state::kb_led_w));
-	map(0x8300, 0x830d).r(FUNC(hp2640_base_state::kb_r));
-	map(0x830e, 0x830e).r(FUNC(hp2640_base_state::switches_ah_r));
-	map(0x830f, 0x830f).r(FUNC(hp2640_base_state::datacomm_sw_r));
-	map(0x8320, 0x8320).w(FUNC(hp2640_base_state::kb_prev_w));
-	map(0x8380, 0x8380).rw(FUNC(hp2640_base_state::switches_jr_r), FUNC(hp2640_base_state::kb_reset_w));
-	map(0x83a0, 0x83a0).r(FUNC(hp2640_base_state::switches_sz_r));
-	map(0x8700, 0x8700).w(FUNC(hp2640_base_state::cx_w));
-	map(0x8720, 0x8720).w(FUNC(hp2640_base_state::cy_w));
+	map(0x8000, 0x8fff).m(m_io_bank, FUNC(address_map_bank_device::amap8));
```
Comment (cuavas) [inline src/mame/drivers/hp2640.cpp:1052]: Can the address map bank device be avoided by using the new “views” feature of the memory system?

File: src/mame/machine/hp2640_tape.cpp [hunk 2]
```diff
@@ -0,0 +1,547 @@
...
+	m_tach_latch = false;
+	m_hole_latch = false;
+	m_byte_ready = false;
+	m_irq = true;
+	m_bit_sync = false;
+	m_isf = false;
+	m_gap = true;
+	m_prev_gap = true;
+	m_rip = false;
+	m_current_op = hp_dc100_tape_device::OP_READ;
```
Comment (cuavas) [inline src/mame/machine/hp2640_tape.cpp:285]: At least some of these should probably be registered for save states.

File: src/mame/machine/hp2640_tape.cpp [hunk 3]
```diff
@@ -0,0 +1,547 @@
...
+	if (state && drive == m_selected_drive) {
+		LOG("MOTION\n");
+		start_rd_wr();
+	}
+}
+
+void hp2640_tape_device::rd_bit_w(unsigned drive , int state)
+{
+	if (drive == m_selected_drive) {
+		bool rd_bit = !!state;
```
Comment (cuavas) [inline src/mame/machine/hp2640_tape.cpp:360]: We tend to use `bool(x)` or `x != 0` to do these, I’d rather we don’t get another convention for converting an integer to Boolean proliferating.

File: src/mame/machine/hp2640_tape.h [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	devcb_write_line m_irq_handler;
+	devcb_write_line m_led0_handler;
+	devcb_write_line m_led1_handler;
+
+	// 0 is left hand drive
+	// 1 is right hand drive
+	required_device_array<hp_dc100_tape_device , 2> m_drives;
+
+	// State
+	unsigned m_selected_drive;  // U35-10
```
Comment (cuavas) [inline src/mame/machine/hp2640_tape.h:53]: You should use an explicitly sized type for this so it can be registered for save states portably.

**Review by cuavas — COMMENTED** [review]
This looks pretty good, there are just a few things that I noticed on reviewing this.

**Comment by fulivi** [comment]
Nice feature this memory_view! I hopefully used it correctly.
While I was at it I also added the displaying of the unit name when tape is in motion: this comes handy when system has more than 1 drive (i.e. HP264x & HP9845).

File: src/mame/machine/hp2640_tape.cpp (hunk 2 above)
Comment (fulivi) [reply src/mame/machine/hp2640_tape.cpp:307]: see device_start

File: src/mame/drivers/hp2640.cpp [hunk 5]
```diff
@@ -1012,19 +1046,40 @@ void hp2640_base_state::cpu_mem_map(address_map &map)
...
+	m_io_view[ 1 ](0x0700, 0x0700).w(FUNC(hp2640_base_state::cx_w));
+	m_io_view[ 0 ](0x0720, 0x0720).w(FUNC(hp2640_base_state::cy_w));
+	m_io_view[ 1 ](0x0720, 0x0720).w(FUNC(hp2640_base_state::cy_w));
+	m_io_view[ 0 ](0x0b00, 0x0b00).w(m_tapes, FUNC(hp2640_tape_device::command_w));
+	m_io_view[ 1 ](0x0b00, 0x0b00).w(m_tapes, FUNC(hp2640_tape_device::command_w));
+	m_io_view[ 0 ](0x0b00, 0x0b00).r(m_tapes, FUNC(hp2640_tape_device::status_r));
+	m_io_view[ 0 ](0x0b20, 0x0b20).w(m_tapes, FUNC(hp2640_tape_device::data_w));
+	m_io_view[ 1 ](0x0b20, 0x0b20).w(m_tapes, FUNC(hp2640_tape_device::data_w));
+	m_io_view[ 0 ](0x0b20, 0x0b20).r(m_tapes, FUNC(hp2640_tape_device::data_r));
+	m_io_view[ 1 ](0x0000, 0x0fff).r(FUNC(hp2640_base_state::poll_r));
```
Comment (cuavas) [inline src/mame/drivers/hp2640.cpp:1081]: If I understand it correctly, you can map the things that don’t change directly into the memory map.  Only the things that differ need to go into the view.  Accesses that don’t match anything in the view should fall through to whatever is in the underlying map.  (@galibert should be able to confirm).

**Comment by galibert** [comment]
Yes, I confirm :-)

  OG.

**Comment by fulivi** [comment]
Is this what you mean?

**Comment by cuavas** [comment]
Yeah, that’s easier to read.

**Pull request merged**